source/fe/mapping_q_cache.cc

```diff
@@ -125,6 +184,340 @@ MappingQCache<dim, spacedim>::initialize(
+      if (vector_describes_relative_displacement || (is_relevant_cell == false))
+        {
+          if (mapping_q_generic != nullptr &&
+              this->get_degree() == mapping_q_generic->get_degree())
+            result =
+              mapping_q_generic->compute_mapping_support_points(cell_tria);
+          else
+            result = fe_values_all.get()->get_quadrature_points();
+
+          if (is_relevant_cell == false)
+            return result;
+        }
+      else
+        {
+          result.resize(
+            Utilities::pow<unsigned int>(this->get_degree() + 1, dim));
+        }
+
+      // Step 2c) read global vector and adjust points accordingly
+      if ((is_fe_q || is_fe_dgq) && fe.degree == this->get_degree())
+        {
+          // case 1: FE_Q or FE_DGQ with same degree as this class has; this
+          // is the simple case since no interpolation is needed
+          std::vector<types::global_dof_index> dof_indices(
+            fe.n_dofs_per_cell());
+          cell->get_dof_indices(dof_indices);
+
+          for (unsigned int i = 0; i < dof_indices.size(); ++i)
+            {
+              const auto id = fe.system_to_component_index(i);
+
+              if (is_fe_q)
+                {
+                  // case 1a: FE_Q
+                  if (vector_describes_relative_displacement)
+                    result[id.second][id.first] +=
+                      vector_ghosted(dof_indices[i]);
+                  else
+                    result[id.second][id.first] =
+                      vector_ghosted(dof_indices[i]);
+                }
+              else
+                {
+                  // case 1b: FE_DGQ
+                  if (vector_describes_relative_displacement)
+                    result[lexicographic_to_hierarchic_numbering[id.second]]
+                          [id.first] += vector_ghosted(dof_indices[i]);
+                  else
+                    result[lexicographic_to_hierarchic_numbering[id.second]]
+                          [id.first] = vector_ghosted(dof_indices[i]);
+                }
+            }
+        }
+      else
+        {
+          // case 2: general case; interpolation is needed
+          // note: the following code could be optimized for tensor-product
+          // elements via application of sum factorization as is done on
+          // MatrixFree/FEEvaluation
+          auto &fe_values = fe_values_all.get();
+
+          std::vector<Vector<typename VectorType::value_type>> values(
+            fe_values->n_quadrature_points,
+            Vector<typename VectorType::value_type>(spacedim));
+
+          fe_values->get_function_values(vector_ghosted, values);
+
+          for (unsigned int q = 0; q < fe_values->n_quadrature_points; ++q)
+            for (unsigned int c = 0; c < spacedim; ++c)
+              if (vector_describes_relative_displacement)
+                result[q][c] += values[q][c];
+              else
+                result[q][c] = values[q][c];
```

Comment:
@kronbichler Regarding the discussion in https://github.com/dealii/dealii/pull/11867#discussion_r592198417, this would be the draft of the fall-back code that needs `FEValues`. Do you have a better idea here?